add implementation of _ProductGeometry

amal-ghamdi opened the pull request:
closes #592 
closes #180

nabriis commented on cuqi/experimental/geometry/_productgeometry.py:
```diff
@@ -0,0 +1,163 @@
+        
+    
+    def __repr__(self) -> str:
+        """Representation of the product geometry."""
```
Printing looks a little off?
![image](https://github.com/user-attachments/assets/d11af2b4-8dd5-4744-84e4-7ffbfa155793)
not sure we need this "of the following geometries"

amal-ghamdi replied:
done 👍

nabriis commented on cuqi/experimental/geometry/_productgeometry.py:
```diff
@@ -0,0 +1,163 @@
+    Parameters
+    ----------
+    geometries : list of cuqi.geometry.Geometry
+        List of geometries to be combined into a product geometry.
```
It seems it should not be a list, but rather comma-seperated arguments.

amal-ghamdi replied:
Thank you, updated 👍

nabriis commented on cuqi/experimental/geometry/_productgeometry.py:
```diff
@@ -0,0 +1,163 @@
+    """
+
+    def __init__(self, *geometries):
+        self.geometries = geometries
```
I suggest adding some error handling to check the input is Geometry

amal-ghamdi replied:
Error handling added 👍

nabriis commented on cuqi/experimental/geometry/_productgeometry.py:
```diff
@@ -0,0 +1,163 @@
+        return sum(self.par_dim_list)
+    
+    @property
+    def par_dim_list(self):
```
Maybe we make this one private?

amal-ghamdi replied:
Since it is documented and not editable and might be used by other classes in the future, I would like to keep it public for now

nabriis commented on cuqi/experimental/geometry/_productgeometry.py:
```diff
@@ -0,0 +1,163 @@
+        return [g.par_dim for g in self.geometries]
+    
+    @property
+    def stacked_par_split_indices(self):
```
Also consider making this private

amal-ghamdi replied:
Since it is documented and not editable and might be used by other classes in the future, I would like to keep it public for now

nabriis commented on cuqi/experimental/geometry/_productgeometry.py:
```diff
@@ -0,0 +1,163 @@
+        raise NotImplementedError(
+            f"Plotting not implemented for {self.__class__.__name__}.")
+
+    def par2fun(self, *args):
```
I don't recall if we support args and kwargs normally or only args?

nabriis commented on cuqi/experimental/geometry/_productgeometry.py:
```diff
@@ -0,0 +1,163 @@
+            f"Plotting not implemented for {self.__class__.__name__}.")
+
+    def par2fun(self, *args):
+        """Converts parameter vector(s) into function values for each 
```
I suggest to add something about the ordering of the geometries being important here.

amal-ghamdi replied:
good suggestion, done

nabriis reviewed (changes requested):
Thanks @amal-ghamdi

Looks good. I had some minor comments.

chaozg commented on cuqi/experimental/geometry/_productgeometry.py:
```diff
@@ -0,0 +1,163 @@
+
+class _ProductGeometry(Geometry):
+    """ A class for representing a product geometry. A product geometry
+    represents the product space of multiple geometries of type Geometry.
```
```suggestion
    represents the product space of multiple geometries of type :class:`Geometry`.
```

chaozg commented on cuqi/experimental/geometry/_productgeometry.py:
```diff
@@ -0,0 +1,163 @@
+            args = self._split_par(args[0])
+
+        # Convert parameter vectors to function values for each geometry.
+        fun_list = []
+        for i, g in enumerate(self.geometries):
+            fun_i = g.par2fun(args[i])
+            fun_list.append(fun_i)
+        return tuple(fun_list)
```
What about `funvals` to keep consistent with `fun2par`?
```suggestion
        funvals = []
        for i, g in enumerate(self.geometries):
            funval_i = g.par2fun(args[i])
            funvals.append(funval_i)
        return tuple(funvals)
```

chaozg commented on cuqi/experimental/geometry/_productgeometry.py:
```diff
@@ -0,0 +1,163 @@
+
+        return  stacked_val if stacked else tuple(par_vec)
+
+    def vec2fun(self, *funvec):
+        """Maps function vector representation, if available, to function 
+        values."""
+        fun_values = []  
```
A tiny reminder that the naming conventions are different here: `funvec` and `fun_values`. Just a reminder and it's up to you if you want to keep them consistent : )

amal-ghamdi replied:
good suggestion, done

chaozg commented on cuqi/experimental/geometry/_productgeometry.py:
```diff
@@ -0,0 +1,163 @@
+
+        return tuple(fun_values)
+    
+    def fun2vec(self, fun):
+        """Maps function values to a vector representation of the function values, if available."""
+        vec_values = []
+        for i, g in enumerate(self.geometries):
+            vec_values.append(g.fun2vec(fun[i]))
```
Also naming convection here. And even tinier reminder: `fun` or `funs` to indicate that it's collection of functions?

amal-ghamdi replied:
Thanks, paying attention to consistency also helped fix the input for fun2vc. I needed to add * there.

chaozg reviewed (approved):
Thank you @amal-ghamdi for this very well-documented and -tested PR. I only left some tiny reminders about naming convection.

amal-ghamdi commented:
Thank you @nabriis and @chaozg for your helpful review. I addressed your comments (@nabriis I am requesting a re-review). I also added a fix for the documentation.

nabriis reviewed (approved):
Thanks @amal-ghamdi. The updates LGTM :)
